src/ray/core_worker/actor_manager.cc

```diff
@@ -13,24 +13,235 @@
 // limitations under the License.
 
 #include "ray/core_worker/actor_manager.h"
+
 #include "ray/gcs/pb_util.h"
 #include "ray/gcs/redis_accessor.h"
 
 namespace ray {
 
-void ActorManager::PublishTerminatedActor(const TaskSpecification &actor_creation_task) {
-  auto actor_id = actor_creation_task.ActorCreationId();
-  auto data = gcs::CreateActorTableData(actor_creation_task, rpc::Address(),
-                                        rpc::ActorTableData::DEAD, 0);
-
-  auto update_callback = [actor_id](Status status) {
-    if (!status.ok()) {
-      // Only one node at a time should succeed at creating or updating the actor.
-      RAY_LOG(ERROR) << "Failed to update state to DEAD for actor " << actor_id
-                     << ", error: " << status.ToString();
-    }
-  };
-  RAY_CHECK_OK(actor_accessor_.AsyncRegister(data, update_callback));
+ActorID ActorManager::DeserializeAndRegisterActorHandle(
+    const std::string &serialized, const ObjectID &outer_object_id,
+    const TaskID &caller_id, const std::string &call_site,
+    const rpc::Address &caller_address) {
+  std::shared_ptr<ActorHandle> actor_handle(new ActorHandle(serialized));
+  const auto actor_id = actor_handle->GetActorID();
+  const auto owner_id = actor_handle->GetOwnerId();
+  const auto owner_address = actor_handle->GetOwnerAddress();
+
+  RAY_UNUSED(AddActorHandle(std::move(actor_handle), /*is_owner_handle=*/false, caller_id,
+                            call_site, caller_address));
+
+  ObjectID actor_handle_id = ObjectID::ForActorHandle(actor_id);
+  reference_counter_->AddBorrowedObject(actor_handle_id, outer_object_id, owner_id,
+                                        owner_address);
+
+  return actor_id;
+}
+
+Status ActorManager::SerializeActorHandle(const ActorID &actor_id, std::string *output,
+                                          ObjectID *actor_handle_id) const {
+  ActorHandle *actor_handle = nullptr;
+  auto status = GetActorHandle(actor_id, &actor_handle);
+  if (status.ok()) {
+    actor_handle->Serialize(output);
+    *actor_handle_id = ObjectID::ForActorHandle(actor_id);
+  }
+  return status;
+}
+
+Status ActorManager::GetActorHandle(const ActorID &actor_id,
+                                    ActorHandle **actor_handle) const {
+  absl::MutexLock lock(&mutex_);
+  auto it = actor_handles_.find(actor_id);
+  if (it == actor_handles_.end()) {
+    return Status::Invalid("Handle for actor does not exist");
+  }
+  *actor_handle = it->second.get();
```

Comment:
Actually, I found 10~20% performance regression when I used shared_ptr. (I ran microbenchmark). I will probably keep sticking to `unique_ptr` for this. But, I will still try to make it return the reference to the pointer.

Reply:
Based on my google search, I think `std::unique_ptr<ActorHandle> blah` return type is fine (without reference) because of compiler Return Value Optimization? I think I just need to do `std::unique_ptr<ActorHandle> &handle = actor_manager_->GetActorHandle()...`

Reply:
Btw, I also found from Google search (I did lots of search haha..) that I should not use the reference of smart pointer unless I am trying to modify smart pointers themselves, and I just need to use raw_pointer if I want to just access the object pointed by smart pointers. If it is true, I think just keeping it makes sense? I am not a cpp expert, so I am not sure if that's what we want though. Let me know your thoughts about this!

Reply:
Hmm that is surprising that there's a 10-20% regression with shared_ptr. One thing is that the microbenchmarks may not be reliable on a laptop (I've seen much greater variation than 10-20% across different runs of the same code). Also, if you use `shared_ptr &` instead of `shared_ptr`, it should not be any different from this method, I think.

By the way, you can't convert a shared_ptr to a unique_ptr or vice versa. Both of them are different ways to manage memory with compile-time safety checks, compared to just calling `new` and `delete` on a raw pointer. The code you have right now is just accessing the raw pointer address that's managed by the shared_ptr. This is not necessarily bad, but we try to avoid it because the caller could then store the raw pointer somewhere else, but then the address may not be valid later on. This is different from a `shared_ptr &`; if the caller wants to store that, it has to copy it to a new `shared_ptr` so we still get memory safety in that case.

Reply:
Yeah it was a bit surprising to me as well because I used `shared_ptr` in the exact same way as `unique_ptr`, and it shouldn't have any overhead theoretically. Weirdly, when I reverted `shared_ptr` to `unique_ptr`,  I got the original result back. (which is super weird!) I also think it is possible that the problem is just because of laptop, but I will keep running more experiment. 

Also, thanks for the information about the pointers! I did notice when I used `unique_ptr`, my unit tests failed because I used `shared_ptr`. One interesting thing I found was that conversion from `unique_ptr` to `shared_ptr` didn't break (although the other way around broke). 

Anyhow, I will then just stick to the solution we return `shared_ptr` here!!